src/nomad_simulations/schema_packages/model_system.py

```diff
@@ -1150,46 +1303,94 @@ def resolve_system_type_and_dimensionality(
 
         return system_type, dimensionality
 
+    def populate_atom_types(self, logger: 'BoundLogger') -> None:
+        """
+        Populates the `atom_types` subsection with unique AtomDefn entries
+        extracted from the `atom_states` subsection.
+        """
+        unique_types = {}
+        for state in self.atom_states:
+            atom_def = state.atom_definition_ref
+            if atom_def is None:
+                logger.warning('Missing atom_definition_ref in one of the atom_states.')
```

Comment:
this could result in a lot of warning, i suggest, putting the warning outside the loop if warning is found